### ISSUE TITLE
Recompiles and cache_size_limit from detectron2 CycleBatchNormList

### ISSUE THREAD

**Issue by jansel** [pytorch/pytorch#122578]
Running
```
./benchmarks/dynamo/torchbench.py --inference --performance --only detectron2_fcos_r_50_fpn --stats --dynamic-shapes
```
I am seeing
```py
WARNING:torch._dynamo.convert_frame:torch._dynamo hit config.cache_size_limit (8)
   function: 'forward' (/home/jansel/conda/envs/pytorch/lib/python3.11/site-packages/detectron2/layers/batch_norm.py:318)
   last reason: L['self']._pos == 0      
```

The root cause is this class (slightly minified by hand):
``
class CycleBatchNormList(nn.ModuleList):
    def __init__(self, length: int, bn_class=nn.BatchNorm2d, **kwargs):
        super().__init__([bn_class(**kwargs, affine=False) for k in range(length)])
        self._pos = 0

    def forward(self, x):
        ret = self[self._pos](x)
        self._pos = (self._pos + 1) % len(self)
        return ret
```

TorchDynamo is correctly detecting the mutation of `self._pos` and mapping the class to `UnspecializedNNModuleVariable`, however the `self[self._pos]` is causing issues because it results in different weights being used each iteration.

This is a particularly tricky case to handle, since it is hard to prove that all values of `self._pos` will result in the same graph.  We also do not support indirect guards, so it is not possible to have a guard on `self[self._pos].blah`.

As a concrete action item, I think we should detect this type of pattern and graph break on it.  This would be a lot better than recompiling and hitting the cache limit, since it would allow us to optimize other parts of the model.
